extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java

```diff
@@ -0,0 +1,382 @@
+        Mutiny.Query<?> jpaQuery = createBaseQuery(em);
+
+        if (range != null) {
+            jpaQuery.setFirstResult(range.getStartIndex());
+            // range is 0 based, so we add 1
+            jpaQuery.setMaxResults(range.getLastIndex() - range.getStartIndex() + 1);
+        } else if (page != null) {
+            jpaQuery.setFirstResult(page.index * page.size);
+            jpaQuery.setMaxResults(page.size);
+        } else {
+            // Use deprecated API in org.hibernate.Query that will be moved to org.hibernate.query.Query on Hibernate 6.0
+            // FIXME: requires Hibernate support
+            //            @SuppressWarnings("deprecation")
+            //            RowSelection options = jpaQuery.unwrap(org.hibernate.query.Query.class).getQueryOptions();
+            //            options.setFirstRow(null);
+            //            options.setMaxRows(null);
+            // FIXME: why would we even do that? those are the defaults, let's leave them blank
+            // if we don't, we get a LIMIT
+            //            jpaQuery.setFirstResult(0);
+            //            jpaQuery.setMaxResults(Integer.MAX_VALUE);
+        }
+
+        return jpaQuery;
+    }
+
+    private Mutiny.Query<?> createQuery(Mutiny.Session em, int maxResults) {
+        Mutiny.Query<?> jpaQuery = createBaseQuery(em);
+
+        if (range != null) {
+            jpaQuery.setFirstResult(range.getStartIndex());
+        } else if (page != null) {
+            jpaQuery.setFirstResult(page.index * page.size);
+        } else {
+            // Use deprecated API in org.hibernate.Query that will be moved to org.hibernate.query.Query on Hibernate 6.0
+            // FIXME: requires Hibernate support
+            //            @SuppressWarnings("deprecation")
+            //            RowSelection options = jpaQuery.unwrap(org.hibernate.query.Query.class).getQueryOptions();
+            //            options.setFirstRow(null);
+            jpaQuery.setFirstResult(0);
+        }
+        jpaQuery.setMaxResults(maxResults);
+
+        return jpaQuery;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Mutiny.Query<?> createBaseQuery(Mutiny.Session em) {
+        Mutiny.Query<?> jpaQuery;
+        if (AbstractJpaOperations.isNamedQuery(query)) {
+            String namedQuery = query.substring(1);
+            jpaQuery = em.createNamedQuery(namedQuery);
+        } else {
+            jpaQuery = em.createQuery(orderBy != null ? query + orderBy : query);
+        }
+
+        if (paramsArrayOrMap instanceof Map) {
+            AbstractJpaOperations.bindParameters(jpaQuery, (Map<String, Object>) paramsArrayOrMap);
+        } else {
+            AbstractJpaOperations.bindParameters(jpaQuery, (Object[]) paramsArrayOrMap);
+        }
+
+        if (this.lockModeType != null) {
+            jpaQuery.setLockMode(LockModeConverter.convertToLockMode(lockModeType));
+        }
+
+        if (hints != null) {
```

Comment:
Alternatively, we should consider simply not supporting hints at all in Panache. They're really there only to allow use of vendor-specific features via the JPA-standard APIs, but they're untypesafe and inconvenient.

I would argue that it would be better if Panache offered typesafe APIs for Hibernate-specific features. As far as I understand, support for other JPA implementations is a non-goal.

Reply:
They will be, let's throw.

Reply:
> They will be, let's throw.

Seriously? you're going to make, say, Eclipselink work in Quarkus?

Reply:
No, I meant hints will be supported, because I thought they would (and github didn't show me your answer at the time, so I was replying to Loïc).

I don't mind if we replace hint support if you tell me HR will not support them.

Reply:
I'd love to fully eradicate support for hints from ORM, and have a proper API instead.

Of course we can't because of specs, but we could introduce an internal configurtion setting to at least ignore them for Hibernate Reactive (and maybe even for ORM classic when within Quarkus).
Beyond API benefits of more explicit options, dealing with them is also highl inefficient.

Reply:
> I don't mind if we replace hint support if you tell me HR will not support them.

Yeah but I'm advocating removing them from the regular (non-reactive) API as well. Or rather, representing the specific hints that Hibernate supports in a type safe way, just like what I've done in certain specific cases in the HR API.